workflows/testnet/testnet.go

```diff
@@ -59,51 +59,87 @@ func Workflow(ctx workflow.Context, req messages.TestnetWorkflowRequest) (messag
 		return "", err
 	}
 
-	buildResult, err := buildImage(ctx, req)
+	buildResult, err := buildImageAndReport(ctx, req, report)
+	if err != nil {
+		return "", err
+	}
 
+	chainState, providerState, err := launchTestnetInternal(ctx, req, runName, buildResult, report)
 	if err != nil {
 		return "", err
 	}
 
-	if err := report.SetBuildResult(ctx, buildResult); err != nil {
+	loadTestRuntime, err := runLoadTestInternal(ctx, req, chainState, providerState, report)
+	if err != nil {
+		workflow.GetLogger(ctx).Error("Load test initiation failed", zap.Error(err))
+	}
+
+	testnetRuntime := 1 * time.Hour // default runtime to 1 hour
+	if req.TestnetDuration != 0 {
+		testnetRuntime = req.TestnetDuration
+	}
+	if loadTestRuntime > testnetRuntime {
+		testnetRuntime = loadTestRuntime
+	}
+
+	// Teardown the provider after monitoring is complete (unless it's a long-running testnet)
+	if !req.LongRunningTestnet {
+		defer func() {
+			err := workflow.ExecuteActivity(ctx, testnetActivities.TeardownProvider, messages.TeardownProviderRequest{
+				RunnerType:    req.RunnerType,
+				ProviderState: providerState,
+			}).Get(ctx, nil)
+			if err != nil {
+				// Log error but don't fail the workflow, as monitoring succeeded
+				workflow.GetLogger(ctx).Error("Failed to teardown provider after monitoring", "error", err)
+			}
+		}()
+	}
+
+	if err := monitorTestnet(ctx, chainState, providerState, req.RunnerType, report, testnetRuntime, req.LongRunningTestnet); err != nil {
+		_ = report.Conclude(ctx, "failed", "error", fmt.Sprintf("Testnet monitoring failed: %s", err.Error()))
 		return "", err
 	}
 
-	var providerState, chainState []byte
-	var providerSpecificOptions map[string]string
+	if err := report.Conclude(ctx, "completed", "success", "Testnet bake completed"); err != nil {
+		return "", err
+	}
 
-	if req.RunnerType == testnettypes.DigitalOcean {
-		providerSpecificOptions = map[string]string{
-			"region":   "ams3",
-			"image_id": "185517855",
-			"size":     "s-4vcpu-8gb",
-		}
+	return "", nil
+}
+
+func buildImageAndReport(ctx workflow.Context, req messages.TestnetWorkflowRequest, report *Report) (messages.BuildDockerImageResponse, error) {
+	buildResult, err := buildImage(ctx, req)
+	if err != nil {
+		_ = report.Conclude(ctx, "failed", "error", fmt.Sprintf("Image build failed: %s", err.Error()))
+		return messages.BuildDockerImageResponse{}, err
+	}
+
+	if err := report.SetBuildResult(ctx, buildResult); err != nil {
+		workflow.GetLogger(ctx).Error("Failed to set build result in report", zap.Error(err))
 	}
 
+	return buildResult, nil
+}
+
+func launchTestnetInternal(ctx workflow.Context, req messages.TestnetWorkflowRequest, runName string, buildResult messages.BuildDockerImageResponse, report *Report) ([]byte, []byte, error) {
```

Comment:
why did you call them internal? they're going to be internal anyways since it starts with a lowercase letter (scoped to package instead of public)